units: HTML table of equivalents within Jupyter notebooks

sadielbartholomew opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/master/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/master/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

This pull request is to address the representation of equivalent units in Jupyter notebook, such that if there are any they are output via a new method ``_repr_html_()``  as a HTML table (or if not, a paragraph is output stating "There are no equivalent units"), where previously the ``repr()`` was instead called, printing a formatted string.

For example (compare this new screenshot to the screenshot from the original Issue as below):

![astropy_final_scr](https://user-images.githubusercontent.com/30274190/67644430-3b8b6300-f919-11e9-91cc-b99702358424.png)

Outside of a notebook environment, the ``repr()`` remains as before, except I have changed the pretty-printed output when there are no equivalent units from ``[]`` to a message stating as such, & amended the one test that relied on this accordingly. I think this makes more sense, but I can revert that slight change to ``__repr__`` is others disagree.

Fixes #9381.

sadielbartholomew commented:
I'm adding a new commit to update the changelog now I know the PR number. I'll check the result of the CI tests in the morning & see what, if anything, needs amending pre-review. Thanks.

sadielbartholomew commented:
Hi maintainers, I might need some advice regarding adding my entry to the changelog; I've been assigned to ``v4.1`` but there is no such listing for ``4.1`` in the changelog. What should I do? Start a new ``4.1`` section, perhaps?

bsipocz commented:
@sadielbartholomew - please ignore the changelog status failure. We're in the middle of branching out the new release, and will add the section for the new 4.1 milestone after the procedure is finished, and thus the status will fail until that happens.

sadielbartholomew commented:
Thanks for clarifying, @bisipocz. I added an entry under ``4.0`` in the meantime, but will remove that in the morning.

sadielbartholomew commented:
>  I added an entry under 4.0 in the meantime, but will remove that in the morning.

Done. To make it easier for when someone adds in a changelog entry later, the entry I added, & just removed, that seems a good concise description of this change, is:
```rst
- ``EquivalentUnitsList`` now has a ``_repr_html_`` method to output a HTML
  table on a call to ``find_equivalent_units`` in Jupyter notebooks. [#9495]
```

I have also added in tests for both this new method ``_repr_html_``, & for the old one, ``_repr_``, which was not covered explicitly, at least as far as I could see.

I'll keep an eye on the CI results.

astrofrog commented on astropy/units/core.py:
```diff
@@ -1409,6 +1406,46 @@ def __repr__(self):
+                return blank_table.format(all_rows)
+
+        @staticmethod
+        def process_equivalent_units(equiv_units_data):
```
This should probably be a private method (``_process_equivalent_units``)

sadielbartholomew replied:
Sure, I'll get it changed as such in my next commit. Thanks for the feedback.

astrofrog commented on astropy/units/core.py:
```diff
@@ -1409,6 +1406,46 @@ def __repr__(self):
+                blank_table = '<table style="width:100%">{}</table>'
+                blank_row_container = "<tr>{}</tr>"
+                heading_row_content = "<th>{}</th>" * self.ROW_LEN
+                data_row_content = "<td>{}</td>" * self.ROW_LEN
```
Since ``Table`` objects look nice in Jupyter notebooks, I wonder if it'd be worth trying to just construct a table here and then using that for ``_repr_html_``? (to avoid duplicating the construction of the HTML table). In fact you could use that for both ``__repr__`` and ``_repr_html_``? I wonder if there is a reason this wasn't done before though, so I'll ping @mhvk, @adrn, and @taldcroft in case they know.

taldcroft replied:
:+1: on using `Table`!  One detail is that the default HTML repr has info about the table class and dtypes, and worries about the max number of lines to display.  So you can fix that up with a subclass like so:
```
class MinimalReprTable(Table):
    def _base_repr_(self, *args, **kwargs):
        kwargs['descr_vals'] = []
        kwargs['max_lines'] = -1
        kwargs['show_dtype'] = False
        return super()._base_repr_(*args, **kwargs)

    def __repr__(self):
        return '\n'.join(self.pformat(max_width=-1, max_lines=-1))
```

astrofrog replied:
@taldcroft - ah this raises an interesting point, currently it looks like ``EquivalentUnitList`` inherits from ``List``, and we can't change that inheritance without breaking user code. So I wonder if returning a Table repr will actually confuse things for users who will then expect it to behave as a table. That would argue for keeping it as it is in this PR - looking much better esthetically but different from Table?

taldcroft replied:
I would argue that consistency in the style (CSS)  of outputted tabular information across astropy is definitely a plus, not a minus.  Agreed that having an class where the repr implies a different object type is not a great thing, but I suspect that only experienced users would recognize the style difference in this PR vs. standard Table style and immediately guess that the `EquivalentUnitList` is not a Table.

One comment I would make on this PR is that I don't that 100% width is the right thing if you happen to be working on 27" screen with a full-width browser open.  With that change then the difference between this output and `MinimalReprTable` is fairly subtle.

astrofrog replied:
@taldcroft - ok I see your point, and I agree then that maybe it does indeed make sense to use the Table functionality to build the repr and html repr. We just need to be careful to not include any mention of 'Table' in the output.

@sadielbartholomew - would you be able to try and se if you can leverage the Table repr functionality to avoid duplicating HTML formatting code? No problem if you don't have time of course, but just wanted to check.

pllim replied:
Maybe try not to import Table at the module level? I don't want the circular import issue to creep in later, like what I see going on with Time.

mhvk replied:
Yes, good point. In `units` we should import nothing else of astropy at the module level - it is needed for almost everything else.

sadielbartholomew replied:
Thanks for the input all, & my apologies, I am completely new to the project & did not realise there was already some code for HTML tables I could have used (or though to check for it).

Saying that, I'm getting mixed messages, so please can someone outline is there is a consensus on whether or not to change it to use a ``Table``?

> would you be able to try and se if you can leverage the Table repr functionality to avoid duplicating HTML formatting code? No problem if you don't have time of course, but just wanted to check.

Admittedly I am short on time, so while I am happy to change to an approach using code from ``Table`` if there's consensus for that, I would need to wait until (very) early November to make a large-scale change realistically (I want to complete Hacktober & started late so have three other PRs to finalise). However I'm happy to deal with any conflict resolution that might result in.

I appreciate I am an outsider to the project & trust all your judgements, but from reading your comments it does seem to me like instead using ``Table`` would take as many lines of code, be less clear/comprehensible, & also require an awkward import, & only defining HTML tags is really duplicated, so I don't personally see any reason to change from what I have. (Happy to go with whatever you decide, though!)

mhvk replied:
@EdwardBetts - thanks for your patience. I would be fine with merging as is - it is a nice improvement and we can do the `Table` version later if need be.

taldcroft replied:
Agreed.  Perfect is the enemy of good!

bsipocz commented:
Given the overall happiness and very quick turnaround of this PR, I'm remilestoning it to 4.0.

astrofrog commented:
I agree with merging without changing to Table as this is already an improvement. I think the only changes to make are the one relating to making the method private, and adding the changelog entry to the 4.0 section

sadielbartholomew commented:
> I think the only changes to make are the one relating to making the method private, and adding the changelog entry to the 4.0 section

@astrofrog I've addressed everything specified here in the latest commit. There was also a [comment from @taldcroft about the table CSS width setting](https://github.com/astropy/astropy/pull/9495#discussion_r339501090), so I have included a change to make that 50%, which seems like a good balance: wide enough to be wider than the default for clarity, but not end up extremely wide when a user has a full-screen notebook open, i.e:

![scr_astr_width](https://user-images.githubusercontent.com/30274190/67763922-ad120100-fa40-11e9-96cf-0951c66d73a1.png)

pllim commented:
Shall we merge then? 🍺

bsipocz commented:
Thank you @sadielbartholomew!  Hope to see you around again in the future :)

sadielbartholomew commented:
My pleasure @bsipocz, I'm very happy to have contributed. Thanks everyone for the quick feedback & comments.
